src/topology/homotopy/equiv.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import topology.homotopy.basic
+
+/-!
+
+# Homotopy equivalences between topological spaces
+
+In this file, we define homotopy equivalences between topological spaces `X` and `Y` as a pair of
+functions `f : C(X, Y)` and `g : C(Y, X)` such that `f.comp g` and `g.comp f` are both homotopic
+to `id`.
+
+-/
+
+universes u v w
+
+variables {X : Type u} {Y : Type v} {Z : Type w}
+variables [topological_space X] [topological_space Y] [topological_space Z]
+
+namespace continuous_map
+
+/--
+A homotopy equivalence between topological spaces `X` and `Y` are a pair of functions
+`to_fun : C(X, Y)` and `inv_fun : C(Y, X)` such that `to_fun.comp inv_fun` and `inv_fun.comp to_fun`
+are both homotopic to `id`.
+-/
+@[ext]
+structure homotopy_equiv (X : Type u) (Y : Type v) [topological_space X] [topological_space Y] :=
+(to_fun : C(X, Y))
+(inv_fun : C(Y, X))
+(left_inv : (inv_fun.comp to_fun).homotopic id)
+(right_inv : (to_fun.comp inv_fun).homotopic id)
+
+namespace homotopy_equiv
+
+instance : has_coe_to_fun (homotopy_equiv X Y) (λ _, X → Y) := ⟨λ h, h.to_fun⟩
+
+@[simp]
+lemma to_fun_eq_coe (h : homotopy_equiv X Y) : (h.to_fun : X → Y) = h := rfl
+
+@[continuity]
+lemma continuous (h : homotopy_equiv X Y) : continuous h := h.to_fun.continuous
+
+/--
+Any topological space is homotopy equivalent to itself.
+-/
+def refl (X : Type u) [topological_space X] : homotopy_equiv X X :=
+{ to_fun := id,
+  inv_fun := id,
+  left_inv := homotopic.refl _,
+  right_inv := homotopic.refl _ }
+
+instance : inhabited (homotopy_equiv unit unit) := ⟨refl unit⟩
+
+@[simp]
+lemma coe_refl (X : Type u) [topological_space X] : ⇑(refl X) = id := rfl
+
+/--
+If `X` is homotopy equivalent to `Y`, then `Y` is homotopy equivalent to `X`.
+-/
+def symm (h : homotopy_equiv X Y) : homotopy_equiv Y X :=
+{ to_fun := h.inv_fun,
+  inv_fun := h.to_fun,
+  left_inv := h.right_inv,
+  right_inv := h.left_inv }
+
+@[simp]
+lemma coe_inv_fun (h : homotopy_equiv X Y) : (⇑h.inv_fun : Y → X) = ⇑h.symm := rfl
+
+@[simp]
+lemma refl_symm (X : Type u) [topological_space X] : (refl X).symm = refl X := rfl
+
+/--
+If `X` is homotopy equivalent to `Y`, and `Y` is homotopy equivalent to `Z`, then `X` is homotopy
+equivalent to `Z`.
+-/
+def trans (h₁ : homotopy_equiv X Y) (h₂ : homotopy_equiv Y Z) : homotopy_equiv X Z :=
+{ to_fun := h₂.to_fun.comp h₁.to_fun,
+  inv_fun := h₁.inv_fun.comp h₂.inv_fun,
+  left_inv := begin
+    refine homotopic.trans _ h₁.left_inv,
+    change ((h₁.inv_fun.comp h₂.inv_fun).comp (h₂.to_fun.comp h₁.to_fun)) with
+      h₁.inv_fun.comp ((h₂.inv_fun.comp h₂.to_fun).comp h₁.to_fun),
+    refine homotopic.hcomp _ (homotopic.refl _),
+    refine homotopic.trans ((homotopic.refl _).hcomp h₂.left_inv) _,
+    rw id_comp,
+  end,
```

Comment:
If you introduce
```lean
infix ` ~ₕ `:25 := homotopic
```
then this proof can be written in calc mode:
```lean
    calc (h₁.inv_fun.comp (h₂.inv_fun.comp h₂.to_fun)).comp h₁.to_fun ~ₕ h₁.inv_fun.comp h₁.to_fun :
      homotopic.hcomp (homotopic.refl _)
        ( calc h₁.inv_fun.comp (h₂.inv_fun.comp h₂.to_fun) ~ₕ h₁.inv_fun.comp id :
            homotopic.hcomp h₂.left_inv (homotopic.refl _)
          ... = h₁.inv_fun : h₁.inv_fun.comp_id)
    ... ~ₕ id : h₁.left_inv,
```

Reply:
I've introduced the notation but left the proof as is.

Reply:
I think that `~ₕ` should be used for homotopy of maps (either `Type*`- or `Prop`-valued; not needed for this PR) and `≃ₕ` should be used for homotopy equivalence of spaces. Otherwise LGTM.
bors d+